Starting quote has no ending quote

**Describe the bug**

When you want to commit and you have an uneven number of quotes (") in your commit message lazygit panics with the message `Starting quote has no ending quote`.

**To Reproduce**

Try to commit with a message like `a"` in laygit

**Expected behavior**

Create a commit with the subject `a"` sd;lfj 1

**Screenshots**

```
panic: Starting quote has no ending quote.

goroutine 1 [running]:
github.com/jesseduffield/lazygit/vendor/github.com/mgutz/str.ToArgv(0xc4204aa080, 0x12, 0x1, 0x1, 0xc42001ead0)
        /build/lazygit-f3gtkp/lazygit-0.0~git20180905.291.ce95e67/obj-x86_64-linux-gnu/src/github.com/jesseduffield/lazygit/vendor/github.com/mgutz/str/funcsPZ.go:377 +0x66b
github.com/jesseduffield/lazygit/pkg/commands.(*OSCommand).RunCommandWithOutput(0xc420248600, 0xc4204aa080, 0x12, 0x0, 0x0, 0xe01fe0, 0xc42001eac0)
        /build/lazygit-f3gtkp/lazygit-0.0~git20180905.291.ce95e67/obj-x86_64-linux-gnu/src/github.com/jesseduffield/lazygit/pkg/commands/os.go:52 +0x127
github.com/jesseduffield/lazygit/pkg/commands.(*OSCommand).RunCommand(0xc420248600, 0xc4204aa080, 0x12, 0xc4204a80a8, 0x4)
        /build/lazygit-f3gtkp/lazygit-0.0~git20180905.291.ce95e67/obj-x86_64-linux-gnu/src/github.com/jesseduffield/lazygit/pkg/commands/os.go:61 +0x3f
github.com/jesseduffield/lazygit/pkg/commands.(*GitCommand).Commit(0xc4202b9860, 0xc420348000, 0xc4204a80a2, 0x2, 0x2, 0x2, 0x0)
        /build/lazygit-f3gtkp/lazygit-0.0~git20180905.291.ce95e67/obj-x86_64-linux-gnu/src/github.com/jesseduffield/lazygit/pkg/commands/git.go:303 +0x1a4
github.com/jesseduffield/lazygit/pkg/gui.(*Gui).handleCommitConfirm(0xc4202a2480, 0xc420348000, 0xc42037af00, 0x0, 0xc42027bd38)
        /build/lazygit-f3gtkp/lazygit-0.0~git20180905.291.ce95e67/obj-x86_64-linux-gnu/src/github.com/jesseduffield/lazygit/pkg/gui/commit_message_panel.go:15 +0x63
github.com/jesseduffield/lazygit/pkg/gui.(*Gui).(github.com/jesseduffield/lazygit/pkg/gui.handleCommitConfirm)-fm(0xc420348000, 0xc42037af00, 0xc42027bd38, 0xc42034a000)
        /build/lazygit-f3gtkp/lazygit-0.0~git20180905.291.ce95e67/obj-x86_64-linux-gnu/src/github.com/jesseduffield/lazygit/pkg/gui/keybindings.go:71 +0x3e
github.com/jesseduffield/lazygit/vendor/github.com/jesseduffield/gocui.(*Gui).execKeybinding(0xc420348000, 0xc42037af00, 0xc4202d0c30, 0xc420000201, 0xc42027bd38, 0xc42027bd38)
        /build/lazygit-f3gtkp/lazygit-0.0~git20180905.291.ce95e67/obj-x86_64-linux-gnu/src/github.com/jesseduffield/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:703 +0x3e
github.com/jesseduffield/lazygit/vendor/github.com/jesseduffield/gocui.(*Gui).execKeybindings(0xc420348000, 0xc42037af00, 0xc42027bca0, 0x6d6ebd, 0x0, 0x0)
        /build/lazygit-f3gtkp/lazygit-0.0~git20180905.291.ce95e67/obj-x86_64-linux-gnu/src/github.com/jesseduffield/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:689 +0x10b
github.com/jesseduffield/lazygit/vendor/github.com/jesseduffield/gocui.(*Gui).onKey(0xc420348000, 0xc420509ca0, 0xc4202fc050, 0xc42035e0b0)
        /build/lazygit-f3gtkp/lazygit-0.0~git20180905.291.ce95e67/obj-x86_64-linux-gnu/src/github.com/jesseduffield/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:650 +0x1b3
github.com/jesseduffield/lazygit/vendor/github.com/jesseduffield/gocui.(*Gui).handleEvent(0xc420348000, 0xc42027bca0, 0xc42027bc98, 0x0)
        /build/lazygit-f3gtkp/lazygit-0.0~git20180905.291.ce95e67/obj-x86_64-linux-gnu/src/github.com/jesseduffield/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:431 +0x40
github.com/jesseduffield/lazygit/vendor/github.com/jesseduffield/gocui.(*Gui).MainLoop(0xc420348000, 0xc420348000, 0x0)
        /build/lazygit-f3gtkp/lazygit-0.0~git20180905.291.ce95e67/obj-x86_64-linux-gnu/src/github.com/jesseduffield/lazygit/vendor/github.com/jesseduffield/gocui/gui.go:391 +0x2e4
github.com/jesseduffield/lazygit/pkg/gui.(*Gui).Run(0xc4202a2480, 0x0, 0x0)
        /build/lazygit-f3gtkp/lazygit-0.0~git20180905.291.ce95e67/obj-x86_64-linux-gnu/src/github.com/jesseduffield/lazygit/pkg/gui/gui.go:382 +0x2cb
github.com/jesseduffield/lazygit/pkg/gui.(*Gui).RunWithSubprocesses(0xc4202a2480)
        /build/lazygit-f3gtkp/lazygit-0.0~git20180905.291.ce95e67/obj-x86_64-linux-gnu/src/github.com/jesseduffield/lazygit/pkg/gui/gui.go:391 +0x2f
main.main()
        /build/lazygit-f3gtkp/lazygit-0.0~git20180905.291.ce95e67/obj-x86_64-linux-gnu/src/github.com/jesseduffield/lazygit/main.go:52 +0x13c
```

**Desktop (please complete the following information):**
 - OS: Linux neon.kde.org
 - Lazygit Version commit=ce95e67, build date=2018-09-06, build source=unknown, version=, os=linux, arch=amd64

--- a/pkg/commands/os.go
+++ b/pkg/commands/os.go
@@ -2,7 +2,6 @@ package commands
 
 import (
 	"errors"
-	"fmt"
 	"os"
 	"os/exec"
 	"strings"
@@ -140,7 +139,6 @@ func (c *OSCommand) PrepareSubProcess(cmdName string, commandArgs ...string) *ex
 
 // Quote wraps a message in platform-specific quotation marks
 func (c *OSCommand) Quote(message string) string {
-	fmt.Println(c.Platform.os)
 	message = strings.Replace(message, "`", "\\`", -1)
 	if c.Platform.os == "linux" {
 		if strings.ContainsRune(message, '\'') {

--- a/pkg/commands/os.go
+++ b/pkg/commands/os.go
@@ -2,6 +2,7 @@ package commands
 
 import (
 	"errors"
+	"fmt"
 	"os"
 	"os/exec"
 	"strings"
@@ -139,6 +140,7 @@ func (c *OSCommand) PrepareSubProcess(cmdName string, commandArgs ...string) *ex
 
 // Quote wraps a message in platform-specific quotation marks
 func (c *OSCommand) Quote(message string) string {
+	fmt.Println(c.Platform.os)
 	message = strings.Replace(message, "`", "\\`", -1)
 	if c.Platform.os == "linux" {
 		if strings.ContainsRune(message, '\'') {

--- a/pkg/commands/os.go
+++ b/pkg/commands/os.go
@@ -17,11 +17,12 @@ import (
 
 // Platform stores the os state
 type Platform struct {
-	os           string
-	shell        string
-	shellArg     string
-	escapedQuote string
-	openCommand  string
+	os                   string
+	shell                string
+	shellArg             string
+	escapedQuote         string
+	openCommand          string
+	fallbackEscapedQuote string
 }
 
 // OSCommand holds all the os commands
@@ -140,14 +141,11 @@ func (c *OSCommand) PrepareSubProcess(cmdName string, commandArgs ...string) *ex
 // Quote wraps a message in platform-specific quotation marks
 func (c *OSCommand) Quote(message string) string {
 	message = strings.Replace(message, "`", "\\`", -1)
-	if c.Platform.os == "linux" {
-		if strings.ContainsRune(message, '\'') {
-			c.Platform.escapedQuote = `"`
-		} else {
-			c.Platform.escapedQuote = `'`
-		}
+	escapedQuote := c.Platform.escapedQuote
+	if strings.Contains(message, c.Platform.escapedQuote) {
+		escapedQuote = c.Platform.fallbackEscapedQuote
 	}
-	return c.Platform.escapedQuote + message + c.Platform.escapedQuote
+	return escapedQuote + message + escapedQuote
 }
 
 // Unquote removes wrapping quotations marks if they are present

--- a/pkg/commands/os.go
+++ b/pkg/commands/os.go
@@ -140,6 +140,13 @@ func (c *OSCommand) PrepareSubProcess(cmdName string, commandArgs ...string) *ex
 // Quote wraps a message in platform-specific quotation marks
 func (c *OSCommand) Quote(message string) string {
 	message = strings.Replace(message, "`", "\\`", -1)
+	if c.Platform.os == "linux" {
+		if strings.ContainsRune(message, '\'') {
+			c.Platform.escapedQuote = `"`
+		} else {
+			c.Platform.escapedQuote = `'`
+		}
+	}
 	return c.Platform.escapedQuote + message + c.Platform.escapedQuote
 }
 

--- a/pkg/commands/os_default_platform.go
+++ b/pkg/commands/os_default_platform.go
@@ -8,10 +8,11 @@ import (
 
 func getPlatform() *Platform {
 	return &Platform{
-		os:           runtime.GOOS,
-		shell:        "bash",
-		shellArg:     "-c",
-		escapedQuote: "\"",
-		openCommand:  "open {{filename}}",
+		os:                   runtime.GOOS,
+		shell:                "bash",
+		shellArg:             "-c",
+		escapedQuote:         "'",
+		openCommand:          "open {{filename}}",
+		fallbackEscapedQuote: "\"",
 	}
 }

--- a/pkg/commands/os_test.go
+++ b/pkg/commands/os_test.go
@@ -265,6 +265,30 @@ func TestOSCommandQuote(t *testing.T) {
 	assert.EqualValues(t, expected, actual)
 }
 
+func TestOSCommandQuoteSingleQuote(t *testing.T) {
+	osCommand := newDummyOSCommand()
+
+	osCommand.Platform.os = "linux"
+
+	actual := osCommand.Quote("hello 'test'")
+
+	expected := osCommand.Platform.escapedQuote + "hello 'test'" + osCommand.Platform.escapedQuote
+
+	assert.EqualValues(t, expected, actual)
+}
+
+func TestOSCommandQuoteDoubleQuote(t *testing.T) {
+	osCommand := newDummyOSCommand()
+
+	osCommand.Platform.os = "linux"
+
+	actual := osCommand.Quote(`hello "test"`)
+
+	expected := osCommand.Platform.escapedQuote + "hello \"test\"" + osCommand.Platform.escapedQuote
+
+	assert.EqualValues(t, expected, actual)
+}
+
 func TestOSCommandUnquote(t *testing.T) {
 	osCommand := newDummyOSCommand()
 

--- a/pkg/commands/os_test.go
+++ b/pkg/commands/os_test.go
@@ -265,6 +265,7 @@ func TestOSCommandQuote(t *testing.T) {
 	assert.EqualValues(t, expected, actual)
 }
 
+// TestOSCommandQuoteSingleQuote tests the quote function with ' quotes explicitly for Linux
 func TestOSCommandQuoteSingleQuote(t *testing.T) {
 	osCommand := newDummyOSCommand()
 
@@ -277,6 +278,7 @@ func TestOSCommandQuoteSingleQuote(t *testing.T) {
 	assert.EqualValues(t, expected, actual)
 }
 
+// TestOSCommandQuoteSingleQuote tests the quote function with " quotes explicitly for Linux
 func TestOSCommandQuoteDoubleQuote(t *testing.T) {
 	osCommand := newDummyOSCommand()
 

--- a/pkg/commands/os_test.go
+++ b/pkg/commands/os_test.go
@@ -272,7 +272,7 @@ func TestOSCommandQuoteSingleQuote(t *testing.T) {
 
 	actual := osCommand.Quote("hello 'test'")
 
-	expected := osCommand.Platform.escapedQuote + "hello 'test'" + osCommand.Platform.escapedQuote
+	expected := osCommand.Platform.fallbackEscapedQuote + "hello 'test'" + osCommand.Platform.fallbackEscapedQuote
 
 	assert.EqualValues(t, expected, actual)
 }

--- a/pkg/commands/os_windows.go
+++ b/pkg/commands/os_windows.go
@@ -2,9 +2,10 @@ package commands
 
 func getPlatform() *Platform {
 	return &Platform{
-		os:           "windows",
-		shell:        "cmd",
-		shellArg:     "/c",
-		escapedQuote: `\"`,
+		os:                   "windows",
+		shell:                "cmd",
+		shellArg:             "/c",
+		escapedQuote:         `\"`,
+		fallbackEscapedQuote: "\\'",
 	}
 }

